create wg-deployment

swiftdiaries opened the pull request:
Represent OWNERs from manifests, kfctl as chairs, tech leads.
If you feel you need to be present in the WG, please comment on the PR.

The `wg-control-plane` can be superset of `wg-deployment`. I wanted to keep things very narrow for scope of this working group. And will be subsequently formed.

Related to: #400

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,35 @@
+#### Cross-cutting and Externally Facing Processes
+
+- Cutting releases on both the manifests and kfctl repos
+- Qualifying a Kubeflow release for each platform
```
I would think qualifying Kubeflow releases for each platform would be the platform owners responsibility?

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,35 @@
+
+### Out of scope
+
+- Maintaining or bug fixes to the individual applications themselves.
```
Should we explicitly call out that deciding what is and is not a conformant Kubeflow distribution is not in scope?

swiftdiaries replied:
I'm not so sure about this. Can we revisit this?

jlewi replied:
Can you explain? Why would conformance be in scope given your charter and how is it relevant to maintaining kfctl and manifests?

jlewi commented:
Thanks for creating this @swiftdiaries should we codify the relationship between app owners and the manifests owner?

Per discussion in #400; I'd like to propose the following

* App owners maintain their application specific manifests upstream in their repos
* This WG maintains automation to copy over these manifests
* This WG maintains tests that ensure individual applications meet certain minimal (non-application specific) expectations
   * e.g. testing that the manifests aren't using deprecated kustomize syntax, or no longer supported K8s resource versions
   * e.g. the tests we have [here](https://github.com/kubeflow/manifests/blob/master/tests/test_util.go)

What should be the relationship between this wg and platform owners?

e.g. suppose the notebook App needs some customization (overlay) specific to a platform (Openshift/AWS/GCP/etc...). Where should this live and who should own it? I can think of three options

1. It lives upstream of manifests in the app repo
1. It lives downstream of manifests in a platform specific repository
1. It lives in the manifests repo

I think 1 & 2 are the most preferable and can be decided on a case by case base by the app owners and platform owners. #3 is the least desirable because in that case 3 parties (as opposed to 2 are involved).

swiftdiaries commented:
I mostly agree with things except 

> This WG maintains automation to copy over these manifests

I think that might create issues around having ambiguous ownership for bug triage and fixes. If there's a bug filed who's responsible for following up on it and fixing it? The application owner or the manifests owner?

I think what we can do is prescriptive manifests for each application with shareable tests maintained by manifests owners.

Prescriptive manifests (one approach off the top of my head) - use of non-relative paths in bases so that application kustomizations can be compiled into platform kustomizations via remote (git) kustomization.

This automatically means that any platform specific kustomization is owned by the platform owner and it's the responsibility of the platform owner to ensure that they develop, maintain and fix issues with it.

Shareable tests - Similar to how we made kfctl E2E available across in manifests previously, `wg-deployment` maintains testing for manifests that's usable across Kubeflow repos. This way we can ensure independent and faster development while also ensuring application compatibility in the broader platform.

Now there are again some issues around common infrastructure layers like Istio or Knative. I was thinking about a way in which we want to ensure that platforms, applications have reasonable default versions. And pre-announcing intended support status for these versions for each release. To give application owners to plan out for sun setting APIs.

For example, we've lagged behind in terms of Istio releases. Istio 1.1 (?) and Istio 1.3 is being offered by different platforms. We probably want to prescribe default versions for each release and ensure each platform is compliant.
Possibly start giving out deprecation notices as we move towards future versions (Istio is at 1.6 I think).

knkski commented:
Hey @swiftdiaries,
Happy to see that this WG is shaping up :+1: 
I've been working on making Kubeflow deploy nicely on MicroK8s and would love to help.

yanniszark commented:
Hi @swiftdiaries! Thanks for creating this PR for `wg-deployment`.
I just came back from PTO and trying to sync up on everything.
Since this WG is about manifests and kfctl, two areas I have been contributing heavily in the past and still am, I would like to be included as a chair and tech lead of `wg-deployment`.

swiftdiaries commented:
Hey @knkski 
That's awesome :) Thank you for your work !

swiftdiaries commented:
Hey @yanniszark 
Added ! Thanks for stepping up :)

andreyvelich commented on wgs.yaml:
```diff
@@ -56,6 +56,67 @@ workinggroups:
+  label: area/wg-deployment
+  leadership:
+    chairs:
+    - github: Jeffwan
```
Should we make it right order ?

swiftdiaries replied:
I think the tooling kind of orders by company

andreyvelich commented on wgs.yaml:
```diff
@@ -56,6 +56,67 @@ workinggroups:
+    chairs:
+    - github: Jeffwan
+      name: Jiaxin Shan
+      company: Amazon Web Services
```
Call it AWS to be consistent with [Training WG](https://github.com/kubeflow/community/blob/master/wgs.yaml#L76)

yanniszark commented on wg-list.md:
```diff
@@ -23,6 +23,7 @@ When the need arises, a [new WG can be created](wgs/wg-lifecycle.md)
 | Name | Label | Chairs | Contact | Meetings |
 |------|-------|--------|---------|----------|
 |[AutoML](wg-automl/README.md)|area/wg-automl|* [Andrey Velichkevich](https://github.com/andreyvelich), Cisco<br>* [Ce Gao](https://github.com/gaocegege), Caicloud<br>* [Johnu George](https://github.com/johnugeorge), Cisco<br>|* [Slack](https://kubeflow.slack.com/messages/wg-automl)<br>* [Mailing List](https://groups.google.com/forum/#!forum/kubeflow-discuss)|* Kubeflow AutoML Working Group Meeti ...
+|[Deployment](wg-deployment/README.md)|area/wg-deployment|* [Jiaxin Shan](https://github.com/Jeffwan), Amazon Web Services<br>* [Animesh Singh](https://github.com/animeshsingh), IBM<br>* [Adhita Selvaraj](https://github.com/swiftdiaries), Cisco<br>* [Yannis Zarkadas](https://github.com/yanniszark), Arritko<br>|* [Slack](https://kubeflow.slack.com/messages/wg-deployment)<br>* [Mailing List](https: ...
```
nit: Arritko -> Arrikto

yanniszark commented:
Thanks @swiftdiaries, only a small nit.
Can you change `Arritko` to `Arrikto` everywhere in the PR?

RFMVasconcelos commented on wgs.yaml:
```diff
@@ -56,6 +56,67 @@ workinggroups:
+    time: "08:00"
+    tz: PT (Pacific Time)
+    frequency: biweekly - every other Tuesday
+    archive_url: http://bit.ly/kf-meeting-notes
```
Will the meeting notes be the same as the current community meet?

swiftdiaries replied:
Yeah, at least in the beginning it will be

vpavlin commented:
Hi @swiftdiaries! I would like to be part of the WG since I am part of the team that is making KF run on OpenShift.

RFMVasconcelos commented:
Hi @swiftdiaries, good to see this wg shaping up! 

I'm working on making Kubeflow deployment smooth for our K8s customers, so would like to be involved 🚀

Jeffwan commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,35 @@
+#### Code, Binaries and Services
+
+- kfctl
+- kfdef for each platform
```
kfdef is part of kubeflow/manifest. Trying to understand if this WG maintains manifest of some common components. lIke profile-controller? Or we want other WG owners to update that? Currently, there's no WG taking care of those components

swiftdiaries replied:
My thoughts are to form a `wg-control-plane` with an explicit mandate to own and maintain those components. The `wg-deployment` could then merge onto it or be sunset with the `wg-control-plane` taking over. My concern is that if we commit to everything as part of one WG, then we might not be able to deliver as much.

Jeffwan replied:
@swiftdiaries Agree. I think it's pretty clear now.

jlewi replied:
Is this working group really maintaining `kfdef for each platform`? e.g. AWS, GCP, OpenShift, etc....
Isn't that unbounded scope since the number of platforms could keep growing?

Jeffwan commented:
Thanks @swiftdiaries driving this! Great progress

yanniszark commented:
/lgtm
/approve

swiftdiaries commented:
Hey @vpavlin 
I've added you as a TL+Chair citing the contributions to both manifests and kfctl (operator). Thank you for stepping up :)

SachinVarghese commented:
Hi, @swiftdiaries I have also been working on kubeflow deployment for existing k8s clusters and other integrations. So it would be great to get involved as part of this working group.

jlewi commented on OWNERS_ALIASES:
```diff
@@ -3,6 +3,12 @@ aliases:
     - gaocegege
     - johnugeorge
+  wg-deployment-leads:
+    - Jeffwan
```
Please alphabetize

swiftdiaries replied:
The tool seems to order it in some fashion.

jlewi commented on wg-deployment/README.md:
```diff
@@ -0,0 +1,39 @@
+* Regular WG Meeting (Pacific PM): [Wednesdays at 17:30 PT (Pacific Time)]() (biweekly - every other Tuesday). [Convert to your timezone](http://www.thetimezoneconverter.com/?t=17:30&tz=PT%20%28Pacific%20Time%29).
+  * [Meeting notes and Agenda](http://bit.ly/kf-meeting-notes).
+
+## Organizers
```
What's the purpose of this section? Is this duplicating the leads listed in the YAML file? Are these folks leads?

jlewi commented:
I took another look through this PR. I think there are two questions that come to mind

1. The PR says this WG maintains kubeflow/manifests; should the charter include some specification of the division of responsibilities between application owners, this WG, and platform owners?

1. Per kubeflow/testing#737 does this WG have a long term plan for test and release infrastructure? Are the leads willing to commit to spinning up and maintaining their own infrastructure if needed?

thesuperzapper commented:
Here is link to my proposal for `kubeflow/manifests`, as this PR probably has most of the people who want to see it:
https://github.com/kubeflow/manifests/issues/1554#issuecomment-692459952

mameshini commented:
I'd like to be present as a contributor and maintainer in this WG.  AgileStacks has a lot of experience with deploying Kubeflow on AWS and on-prem Kubernetes.  Here is GitHub repository for AgileStacks Kubeflow distribution on EKS: https://github.com/agilestacks/stack-ml-eks

Please add @mameshini as the point of contact for Agile Stacks distribution:
https://github.com/mameshini), AgileStacks [Igor Mameshin]

swiftdiaries commented:
@mameshini 
Could you please list your contributions to `kubeflow/manifests` or `kubeflow/kfctl`?

swiftdiaries commented:
> 1.   The PR says this WG maintains kubeflow/manifests; should the charter include some specification of the division of responsibilities between application owners, this WG, and platform owners?

Updated charter.

> 2.    Per kubeflow/testing#737 does this WG have a long term plan for test and release infrastructure? Are the leads willing to commit to spinning up and maintaining their own infrastructure if needed?

Updated charter with the responsibilities. The current plan is to use GitHub Actions as an interim while investigating longer term alternatives.

Jeffwan commented:
@swiftdiaries @jlewi 

>> Per kubeflow/testing#737 does this WG have a long term plan for test and release infrastructure? Are the leads willing to commit to spinning up and maintaining their own infrastructure if needed?

> Updated charter with the responsibilities. The current plan is to use GitHub Actions as an interim while investigating longer term alternatives.

I think AWS would like to help on the test-infra side once credits is approved. onprem and AWS manifest can run there. In order to run tests on other platforms, it needs credentials. kfctl and manifest presubmit/periodical jobs are in the scope.

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,51 @@
+### In scope
+
+#### Responsibilities
+
```
Can you add some additional clarification here? I think this reads like all the responsibilities listed in the subsections below are being assumed by this WG.  When in fact its just the opposite. Maybe move this into the cross cutting section and expand on how you intend to work with the individual WG charters?

swiftdiaries replied:
Yeah, that was my intention. Let me change that

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,51 @@
+- Co-ordinating with application OWNERs before a release to qualify application releases
+
+### Out of scope
+
```
Do you want to call out maintaining manifests for individual applications is out of scope for this WG? That this should be the responsibility of the application OWNERs and WGs?

swiftdiaries replied:
Yeah, I've listed it as an responsibility of the application OWNERs. I'll add that in out of scope for this WG

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,51 @@
+
+## Scope
+
+Manage manifests for deploying and operating Kubeflow.
```
I'm wondering if there's a way to reword this to make it clear that deploying and managing individual applications is outside the scope of this WG.

What about something along the lines of

> Provide a catalog (centralized repository) of Kubeflow application manifests
> Provide tooling to deploy Kubeflow applications from the catalog

Does this seem more inline with the proposd division of responsibility between this WG and the other WGs?

swiftdiaries replied:
Yeah that makes things clearer :) Thanks for the review !

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,51 @@
+
+#### Responsibilities
+
+##### Application Owners
```
Under responsibilities:

How about maintaining a test suite to ensure all application manifests satisfy basic requirements? e.g. ensuring they aren't using obsolete features in kustomize, Kubernetes, ISTIO etc... (we already have some minimal tests)?

jlewi commented:
@swiftdiaries and @yanniszark What's the current thinking with respect to the kfctl_k8s_istio.yaml, and kfctl_istio_dex.v1.1.0.yaml configurations. Are those something this WG will own? If so should that be its own project within this WG? What would the scope be.

My current thinking [per above](https://github.com/kubeflow/community/pull/402/files/5bb29323d7421d4dca557945268388eb66235139#diff-338fd856b3909eabf1dabf9fe2204bba) is that this WG has the following projects

* Provide a catalog (centralized repository) of Kubeflow application manifests
   * This covers a large percentage of kubeflow/manifests
* Provide tooling to deploy apps from the catalog (i.e. kfctl)

The remaining bits in kubeflow/manifests are
   * Platform specific configurations with clear OWNERs (e.g. IBM, AWS, GCP)
       * I would expect these to be OWNED by appropriate vendors and not the WG
   * The "generic" K8s configurations
      * The ownership model here is unclear and needs to be resolved

For the "generic" K8s configurations, I can think of two options

1. We make them their own properly scoped project within the WG
1. The WG says it doesn't want to own them in which case either 
     i) someone decides to maintain them outside of KF as an oppinionated distribution (equivalent to AWS, GCP)
    ii)  WG claims them

swiftdiaries commented:
I feel like there's good amount of users for those configurations to warrant the WG to claim them and help in shaping maintaining them.

jlewi commented:
> I feel like there's good amount of users for those configurations to warrant the WG to claim them and help in maintaining them.

I'm supportive provided we can properly scope it.

> Kubeflow deploy configurations on "generic" Kubernetes

Seems a bit unbounded. How many configurations are you intending to support? What scenarios are you targeting; e.g. no cluster? local cluster (e.g kind)? How far down the rabbit hole of installing ISTIO, knative, Dex, etc... are you going to go.

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,48 @@
+This charter adheres to the conventions, roles and organization management outlined in [wg-governance].
+
+## Scope
+
```
```suggestion
* Provide a catalog (centralized repository) of Kubeflow application manifests
* Provide tooling to deploy apps from the catalog (e.g kfctl)
```

swiftdiaries commented:
Cleaned up commits and rebased.

jlewi commented:
For the "generic" K8s configurations; there was a lot of discussion about what to do. 

A couple observations: 

* I don't think there was a lot of consensus on what the objectives/make up of a "generic" configuration should be; I believe there were at least three different ideas put forth

   * Provide a generic configuration that installs only Kubeflow applications but not middleware (e.g Knatie, ISTIO, Dex)
   * Provide Kubeflow applications and middleware
   * Provide different configurations for  dev/training and production/serving

@animeshsingh pointed out other projects (e.g. Tekton) make it easier to install each piece separately.

I think if the scope of this charter is

    * Maintain an application catalog
    * Provide tooling simplify deploying applications from the catalog   

then that is in line with @animeshsingh suggestion of making it easy for people to select which applications they want to install.

That would put the generic configurations outside the scope of the wgs and on equal footing with the other opinionated distributions (e.g. GCP, AWS, IBM). 

To bring scope in line with the status quo; we could add

    * Host a catalog of 3rd party contributed opinionated distributions of Kubeflow

I think this clarifies that the scope of the WG isn't the distributions themselves (e.g. kfctl_k8s_istio) but just providing a repo & directory structure that makes it easy for users to find them. 

What do folks think?

jlewi commented:
@Jeffwan @animeshsingh @yanniszark @swiftdiaries @vpavlin @yanniszark What's the path forward here?

Jeffwan commented:
> Provide a generic configuration that installs only Kubeflow applications but not middleware (e.g Knatie, ISTIO, Dex)

I think currently some core functionality relies on the these libraries. I feel it would be hard to provide a multi-user version without istio and dex. 

`a generic configuration`  could be minimum third-party middlewares + kubeflow applications. Do we have consensus on this ?

I am not sure how many efforts we want to serve an application catalog. Maybe we can use folder structure to reorganize it firstly and gradually improve it towards catalog or other ways Tekton using

> I think this clarifies that the scope of the WG isn't the distributions themselves (e.g. kfctl_k8s_istio) but just providing a repo & directory structure that makes it easy for users to find them.

Agree.

jlewi commented:
@Jeffwan @animeshsingh @yanniszark @swiftdiaries @vpavlin @yanniszark Does someone want to take a stab at updating the PR so we can finalize this?

animeshsingh commented:
_That would put the generic configurations outside the scope of the wgs and on equal footing with the other opinionated distributions (e.g. GCP, AWS, IBM).

To bring scope in line with the status quo; we could add

* Host a catalog of 3rd party contributed opinionated distributions of Kubeflow_

The revised scope makes sense. Additionally we should remove this from the scope

_Testing suite for application catalogs
Infrastructure for testing application catalogs_

The two repos this group owns are **kfctl and manifests**, and the kustomize manifests for different applications come from the respective wgs.

@swiftdiaries checking with you if this is still something you would have bandwidth to drive? 

Additionally would be good to get an updated commitment from others. 

@Jeffwan @yanniszark @vpavlin

swiftdiaries commented:
Hey @animeshsingh 
I'm oversubscribed on commitments at the moment and don't have the cycles to drive this. I am so sorry.

RFMVasconcelos commented:
@jlewi I agree with the view of scope as:
> * Maintain an application catalog
> * Provide tooling to simplify deploying applications from the catalog   
> * Host a catalog of 3rd party contributed opinionated distributions of Kubeflow

once there seems to be multiple 3rd party contributed opinionated distributions of Kubeflow that could benefit the solidity of the overall project given a well organized catalog.

Does this mean that the `wg-deployment` will own an additional repo besides kfctl and manifests, or that the kfctl and/or manifests repo would somehow include these 3rd party methods?

jlewi commented:
@Jeffwan @animeshsingh @yanniszark @swiftdiaries @vpavlin @yanniszark Can one of the other leads please pick up this PR?

@swiftdiaries I think if you give the other leads write permission on your repo they can push commits to your fork/branch to update the PR.

It would be great to not "fork" the PR so we have a single PR with the complete history.

As @animeshsingh points out we need

1. An updated PR with revised scope
1. Signoff by the other leads.

jlewi commented:
/cc @cvenets

swiftdiaries commented:
Yeah sure thing, if one of the leads can volunteer I can give them write access. 
Sorry for making this process harder  l :(

Jeffwan commented:
@swiftdiaries no worries. We are almost there. I can help update the doc. Please grant permission to me on this repo and I will update it based on @animeshsingh's feedback

jlewi commented:
@cvenets and @yanniszark I think we really need to close on the scope of the deployments WG before taking up your proposal for the [control plane](https://docs.google.com/document/d/1BpR4RTp4m2GO_r-c3_xZ-qGnCfvQdqiRh4V0cFUSBqA/edit).

Right now it seems like there is a lot of overlap.

cvenets commented:
@jlewi indeed there is a lot of overlap, due to the manifests repo.

We discussed this internally and think that tying together the kfctl repo with the manifests repo makes sense only for vendors who are planning to use kfctl.

Thus, assuming the Control Plane WG goes forward and we provide a common manifests repo based only on bare kustomize, for everyone to build upon or use it as is, we will remove ourselves from the deployments-wg and let anyone who wants to maintain kfctl take it from there.

jlewi commented:
@cvenets I'm totally confused here. So are you proposing that the deployments working group not maintain manifests anymore? Have you discussed this with the other proposed leads of the deployments WG?

mameshini commented:
What WG can continue to maintain and test kfctl based deployments?  I think
this group - deployment WG is a natural fit for owning kfctl.  kfctl is
currently used to package manifests in a way that is stable and tested
across several platforms and distros.  Using kfctl is easier than using
kustomize, and completely ignoring the issue of composition, packaging,
version compatibility of applications is not going to make Kubeflow more
stable or easier to deploy.  Ignoring these issues will make Kubeflow
unpredictable in deployments.  We can move kfctl into deployment&testing WG
if it's out of scope for Control Plane WG.  I hope kfctl will continue to
be a core part of Kubeflow, and willing to help with ongoing contributions
and testing for kfctl as part of wg-deployment or wg-controlplane, wherever
it lands.

On Sun, Oct 25, 2020 at 9:49 AM Jeremy Lewi <notifications@github.com>
wrote:

> @cvenets <https://github.com/cvenets> I'm totally confused here. So are
> you proposing that the deployments working group not maintain manifests
> anymore? Have you discussed this with the other proposed leads of the
> deployments WG?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/community/pull/402#issuecomment-716177189>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSIPTH5ESZT52A2647QFFLSMRJK5ANCNFSM4QNQ3ZKQ>
> .
>

cvenets commented:
@jlewi I'm completely on-board with the separation of responsibilities, and think it is the right path forward. The app manifests should be the single point of truth and everyone should be starting/syncing from there.

What I was trying to say is that currently the deployments-wg tries to tie together the kfctl repo with a manifests repo. Before the proposal of the Control Plane WG, this made sense, because we assumed that we had people wanting to maintain kfctl-based deployments/tests, and thus needed manifests as well.

What we propose in the Control Plane WG is a set of kustomize manifests that will allow everything to work together, deploy and test. Kustomize is already the tool of choice of the KFP and KFServing WGs, so we think it is the only common denominator. And it makes sense anyways. The way we will also describe it in the PR, is that we keep the separation, we sync the app manifests (single source of truth) to the Control Plane's manifests repo, and then provide more manifests for the common services and any overlays that may be needed in this repo.

If someone wanted to use only the Control Plane manifests, and the corresponding instructions, then yes, they wouldn't need the manifests in deployments-wg. But if someone (e.g., @mameshini)  wanted to use kfctl, they would need to do two things:

1. Maintain the kfctl code somewhere
2. Either create kfctl manifests from scratch, or use/sync the Control Plane's manifests in the deployments-wg repo and add kfctl-specific overlays on top. The latter will be probably easier to do for people that care about kfctl.

I'm guessing, for the above two, it would make sense to do it in the deployments-wg, with its original purpose, if I have understood correctly.

jlewi commented:
>  deployments-wg tries to tie together the kfctl repo with a manifests repo

That is not my understanding. The scope as currently defined is

> * Provide a catalog (centralized repository) of Kubeflow application manifests
> * Provide tooling to deploy apps from the catalog (e.g kfctl)

The only scalable way (and inline with the discussion) to build a catalog is to offload curation of the app manifests to app owners themselves and just aggregate the manifests in the repo. Thus, it follows that the app manifests wouldn't be tightly coupled to kfctl.

I don't think we want to have two application catalogs. So assuming we move forward with the deployments WG (and I hope we will) then I don't think we would approve another WG maintaining a second catalog; so a catalog would be out of scope for a CP-WG.

@Jeffwan @animeshsingh @yanniszark @swiftdiaries @vpavlin @yanniszark  it would be great to finalize this WG. Here's what I think needs to happen

1. @swiftdiaries please grant write access on your repo to the other leads so they can edit the PR
1. Other leads please agree on scope and membership
    * @cvenets please make a decision about whether you want to be a lead or not on this WG
1. Update the PR to reflect all the changes
1. All leads should LGTM the PR to indicate they are on board with the latest revisions
1. @kubeflow/project-steering-group  will review the PR

animeshsingh commented:
@jlewi what I understand from @cvenets is that except for kfctl + vendor specific manifests , the basic manifests catalog working on an OSS core comes out of the control-plane wg. I am onboard with it.

At this point @swiftdiaries is stepping back, @yanniszark represents Arrikto, so that leaves @Jeffwan and @vpavlin to vote one way or other.

cvenets commented:
@jlewi I agree with @animeshsingh. Our proposal is for the base catalog to come out of the CP WG, and be in sync with what app owners publish.

I've reached out to @PatrickXYS from AWS during the weekend and he is happy to co-lead the CP WG. We'd be happy to add @Jeffwan as well if he likes, since he has an established track record of contributions, to help @yanniszark with the re-factoring.

jlewi commented:
This PR has been open for over two months. As of two weeks ago and last weeks community meeting various proposed leads were pushing to finalize the WG. Are we really going back to square one and proposing that one of the core responsibilities of this WG now be split out into a separate WG?

cvenets commented:
@jlewi I don't think we are back to square one, to the contrary.

It's much clearer what needs to be done now, and it seems that most of the community is aligned. I actually think that the plan forward is much closer to your initial thoughts of keeping everything separate, and clearly abstracted. Both separate from the app side, and separate from the opinionated tooling side.

@yanniszark from our team will be putting the detailed proposal out, and I think this will push things forward. I understand that this proposal was open for a long time, and I would also like to see things moving forward faster, but this also depends on the actual effort that people are open to putting into it. I think with the official PR, things will become very clear.

@yanniszark and the rest of our team will be putting a lot of effort the next months to re-factor the manifests to the desired state per almost all your initial suggestions as well and the discussion the last weeks, so I think that the Control Plane WG will help a lot at advancing the current state.

Jeffwan commented:
Adhita gave me permission to modify his PR last night. I will update this one today. Let's see any feedbacks from other WG-leaders.

jlewi commented:
@cvenets and @yanniszark Before you invest time creating a new PR can you please explain why the sudden change of strategy?

Looking at your [doc](https://docs.google.com/document/d/1BpR4RTp4m2GO_r-c3_xZ-qGnCfvQdqiRh4V0cFUSBqA/edit?ts=5f96e397) it looks like what your proposing is exactly what we have been discussing for the past two months and what is in this PR.

Here's what @yanniszark replied in a comment thread

> All in all, a core objective of this WG [control plane wg] is to establish this flow of work from upstream apps to the manifests repo. This work should not be conflated with opinionated deployment tools, like kfctl, which tend to force a specific structure (e.g., params.env files for env substitution) that make manifests unmaintainable and conflate responsibilities between app and platform owners.

Which sounds an awful like what we've been discussing for the past [two months](https://github.com/kubeflow/community/pull/402#issuecomment-682704600)

> App owners maintain their application specific manifests upstream in their repos
> This WG maintains automation to copy over these manifests

We've likewise said oppionated distributions are out of scope for the deployments WG

> That would put the generic configurations outside the scope of the wgs and on equal footing with the other opinionated distributions (e.g. GCP, AWS, IBM)

Should we push pause on forming a deployments and control plane WG for 3-6 months to give the respective parties time to reset?

cvenets commented:
@jlewi I think @yanniszark has completed most of the PR by now, and was planning to share it either today or tomorrow. I'll let him post once he is completely done, due to US/Europe time difference. No worries about the effort.

I don't think it is really a change of strategy. My understanding is that @yanniszark was planning to do the bulk of the development work for re-factoring the manifests repo independently of WG, he has been part of the discussion the last weeks, and this is why the two proposals are so aligned, which is actually a good thing.

The only difference I see, is that we deliberately want to make sure we don't get distracted with kfctl maintenance, and it's left to the interested vendors to do so. And also have the framework to put the additional work to maintain the unmaintained components, their manifests, as well as manifests for the 3rd party common services in a holistic way, per the proposal.

I'm fine with getting feedback from everyone and waiting a bit, but until now my feeling is that we pretty much have consensus at least from the big vendors co-leading the Deployments WG. IBM agrees, AWS is happy to co-lead the CP WG with us and help test (@PatrickXYS or/and @Jeffwan please confirm here as well if you like), Cisco has stepped back from the Deployments WG already, and most people from the community in the relevant email thread seem to think trying to pursue this is a good thing for the Kubeflow project and the broader community.

We are trying our best to help as much as we can, judging from what we've seen until now, and the effort we see that's needed in front of us, so hopefully this is and will be helpful.

jlewi commented:
> My understanding is that @yanniszark was planning to do the bulk of the development work for re-factoring the manifests repo independently of WG,

Why would @yanniszark or anyone from this WG be re-factoring the manifests? Wouldn't it be the app owners and respective WGs that would be responsible for maintaining their manifests?

Jeffwan commented:
- Address "Host a catalog of 3rd party contributed opinionated distributions of Kubeflow" that indicates this WG is not distributions.  @jlewi 
- Remove test related items. @animeshsingh 

It should be good to go

Jeffwan commented:
> @Jeffwan please confirm here as well if you like

I would like to move forward WG-deployment first and see if there's good proposals on new options to distribute Kubeflow at the same time. Most AWS users are still on `kfctl + manifest`.  I think AWS is open to alternatives and @PatrickXYS is willing to put some efforts on evaluation and moving off kfctl. We are not sure if existing AWS specifics logics can be moved to new way. I think this won't be finished soon, it's kind of long term investment as we did before like  shell to kfctl, ksonnet to kustomize, kustomize to v3 stack, etc. 

I am not sure if community should review proposal first or have the WG-control-plane setup first.

jlewi commented:
Thanks @Jeffwan 

I believe the main delta now between @Jeffwan 's proposal and [Arrikto's latest proposal](https://docs.google.com/document/d/1BpR4RTp4m2GO_r-c3_xZ-qGnCfvQdqiRh4V0cFUSBqA/edit?ts=5f96e397) is what to do about the orphaned applications
   * Central Dashboard
   * Profile Controller
   * PodDefaults Controller

I think for each of these we should pursue one of two paths
    
   * Finding an existing WG to sponsor them 
   * Turning them down

* PodDefaults  - This doesn't seem very specific to Kubeflow so I would propose that future development should move outside of Kubeflow

* Central Dashboard - Since Arrikto is proposing to take this on, how about getting the notebooks WG to sponsor this?
    
* Profile Controller - I believe our plan per @yanniszark 's [proposal](https://github.com/kubeflow/community/blob/master/guidelines/auth.md) is to get rid of KFam and Profiles and just rely on K8s / ISTIO RBAC
    * So I believe we only have to continue to maintain this long enough to give the existing applications time to migrate
    * I believe the biggest dependency on profile controller is centraldashboard (I think pipelines might also have some dependency)
    * So would it make sense to sponsor profile controller also in notebooks WG?

RFMVasconcelos commented:
Not wanting to add friction to the creation of a productive WG, I think a few points are still unclear:

Is kfctl part of this WG?
- According to [Arrikto's latest proposal](https://docs.google.com/document/d/1BpR4RTp4m2GO_r-c3_xZ-qGnCfvQdqiRh4V0cFUSBqA/edit?usp=sharing) it is not, according @Jeffwan's proposal it is. 

> Host a catalog of 3rd party contributed opinionated distributions of Kubeflow
- If this is the case, would `kfctl` fit as one of the opinionated distributions? (see [thread](https://groups.google.com/g/kubeflow-discuss/c/d6whgEgror8/m/4lmS8OMmAAAJ)) 
- What other [distributions ](https://docs.google.com/document/d/1HcdDpORO-ed0oJkKcuGg8GKmo8i-rKHot-9PO95e7KM/edit?usp=sharing) would be included?

On our side we also only use `manifests`, but given that `kfctl` is still required by some vendors (AWS, AgileStacks) it does make sense to maintain it somewhere. Would it make sense to include `kfctl` as part of this catalog of 3rd party opinionated distributions?

jlewi commented:
kfctl is not an oppinionated distribution. kfctl is a tool that aids in deploying applications from the catalog; hence it falls within the scope of

> Provide tooling to deploy apps from the catalog

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,44 @@
+This charter adheres to the conventions, roles and organization management outlined in [wg-governance].
+
+## Scope
+* Provide a catalog (centralized repository) of Kubeflow application manifests
```
I believe the catalog is now out of scope because we have a separate WG: #435

Jeffwan replied:
So WG-manifest will also take care of existing manifests? I thought WG-manifest will create a different catalog for next generation without using kfctl to orchestrate.

Jeffwan replied:
I don't WG-deployment won't provide a new one but need to maintain existing one. I will remove this and reorganize the content.

jlewi replied:
Thanks @Jeffwan 

Here's a bit of additional context.

My understanding is that we  are basically splitting the deployments WG into two

* deployments wg - owns kfctl
* manifests wg - owns the catalog

The motivation for splitting them in two is

* Increase accountability for kfctl and catalog by having two WGs with smaller scope and smaller number of leads
* Ensure that catalog is independent of kfctl

/cc @yanniszark @cvenets

animeshsingh commented on wg-deployment/README.md:
```diff
@@ -0,0 +1,39 @@
+* Animesh Singh (**[@animeshsingh](https://github.com/animeshsingh)**), IBM
+* Adhita Selvaraj (**[@swiftdiaries](https://github.com/swiftdiaries)**), Cisco
+* Vaclav Pavlin (**[@vpavlin](https://github.com/vpavlin)**), Red Hat
+* Yannis Zarkadas (**[@yanniszark](https://github.com/yanniszark)**), Arrikto
```
If @swiftdiaries is stepping out - maybe we see if he can suggest a replacement from Cisco, or clarify if Cisco has any stake in it?

Similarly I believe @yanniszark and Arritko are moving away from kfctl

animeshsingh replied:
@mameshini is Agile stack interested in contributing and maintaining kfctl?

@vpavlin we haven't heard any revised information from RedHat? I believe you have. a stake and are interested in contributing and maintaining kfctl?

cvenets replied:
@animeshsingh you are correct. Once the Manifests WG gets accepted, we (Arrikto) will move away from this WG.

mameshini replied:
@animeshsingh Yes AgileStacks engineers will participate in kfctl enhancements and maintenance, with code contributions and testing of kfctl on AWS and on-prem.  I can be the point of contact for development priorities or PR reviews.

Jeffwan replied:
Great. I will make some changes

jlewi replied:
It is not my intention to dissuade anyone from being a lead; I also intend to let the leads self-organize. That said let me try to clearly communicate my expectation.

The expectation that I've tried to set in other WGs is that the leads are responsible for the success or failure of the projects.

This is intentionally & necessarily a higher bar then just contributing. My expectation is that leads have significant skin in the game and are willing and able to invest the resources needed to make their projects successful

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+
+## Scope
+
+WG Deployment is responsible for Kubeflow release and tooling to deploy apps from the catalog (e.g kfctl).
```
nit: What is meant by "Kubeflow release"  here; 
maybe just drop it.

Jeffwan replied:
removed

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+#### Responsibilities
+
+- Provide tooling to deploy Kubeflow applications from the catalog.
+- Work with Working Group / Application Owners and Platform Owners to make the integration and release process easier.
```
Is this item about the release process still relevant?

Jeffwan replied:
removed

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+#### Code, Binaries and Services
+
+- kfctl (including operator)
+- manifests
```
manifests should be removed since this will now be under the manifests WG.

Jeffwan replied:
removed

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+#### Cross-cutting and Externally Facing Processes
+
+##### With Application Owners
+- Co-ordinating with application OWNERs before a release to qualify application releases
```
"Co-ordinating with application owners" ; this no longer seems in scope given the new manifests proposal.

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+- Ensure application manifests are up to date with the latest changes.
+
+##### With Platform Owners
+- Ensure platform-specific deployment artifacts (manifests, KFDefs, kpt functions) are up-to-date.
```
No longer seems relevant

Jeffwan replied:
Added

jlewi commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,43 @@
+- Ensure platform-specific deployment artifacts (manifests, KFDefs, kpt functions) are up-to-date.
+
+### Out of scope
+
```
Lets add a couple things to being out of scope

* Defining or maintaining a Kubeflow distribution
* Deciding which applications or versions of applications should be included in KF distributions

Jeffwan replied:
Added

jlewi commented:
Thanks @Jeffwan, @animeshsingh et Al for helping pushing this over the hump.

I did another pass through and I think we are almost there. I think these are the remaining steps

1. I left a couple nit-picky comments on the charter
1. I'd like an explicit LGTM from all proposed leads for the updated charter
1. Once the above are satisfied I will be ready to LGTM
1. @paveldournov and @theadactyl will also need to LGTM as members of the PSG.

RFMVasconcelos commented:
Hi all! 

Respecting everyone's invested time in this, I am still a bit confused. 

In the last community meeting, opinions seemed to indicate that `kfctl` was going to be phased-out at some point (without consensus I think). Also, there have been many discussions ([email thread](https://groups.google.com/g/kubeflow-discuss/c/d6whgEgror8) and later #434) around the role of a minimal "generic" distribution. 

This charter seems to indicate that `kfctl` will be set as the default "generic" distro. Is this the case?

Bobgy commented on OWNERS_ALIASES:
```diff
@@ -12,6 +12,12 @@ aliases:
+  wg-deployment-leads:
+    - Jeffwan
+    - animeshsingh
+    - swiftdiaries
```
Should @swiftdiaries be removed from the list?

Jeffwan replied:
resolved

jlewi commented:
@RFMVasconcelos 

> This charter seems to indicate that kfctl will be set as the default "generic" distro. Is this the case?

kfctl is not a distribution. It is a porcelain wrapper around kustomize that makes it easy to  deploy KF manifests. 

The scope of the WG is

> WG Deployment is responsible for Kubeflow release and tooling to deploy apps from the catalog (e.g kfctl).

This seems justified in terms of

* Taking ownership of what's already part of Kubeflow (e.g. kfctl) and figuring out the future

A distribution as defined in #434 is

> A Kubeflow distribution is an opinionated bundle of Kubeflow applications optimized for a particular use case or environment.

kfctl is a tool that can be used to define/deploy distributions but it is not a distribution in itself.

Per my most recent comments the following should be explicitly declared as out of scope

> * Defining or maintaining a Kubeflow distribution
> * Deciding which applications or versions of applications should be included in KF distributions

Jeffwan commented:
@jlewi  @cvenets @yanniszark  One thing I am not clear if WG-manifest will keep maintaining existing manifest? From my understand, Arriko's proposal would like to create some manifests different to manage Kubeflow applications? in that case, this group need to take over the `kubeflow/manifest`, if not, we can totally remove any manifest related items in this proposal. 

@animeshsingh @PatrickXYS @mameshini @vpavlin
If the scope is kfctl, do we still want to setup a WG? I feel like owning repo and drive direction there is good enough. 

If WG-deployment will come up some solutions and doesn't need kfctl to orchestrate, then what's the meaning for this WG to "provide tools to deploy applications from the catalog"? kfctl can goes into maintenance mode.

animeshsingh commented:
@Jeffwan I feel we are going in circles. Key is that kfctl exists, and is still the default tooling for Kubeflow deployment across multiple orgs, including AWS, IBM, Agile Stack, RedHat and a long tail of consumers. As part of the WG when it gets formalized we can decide on on a 3 months, 6 months and 1 year roadmap how this evolves, but at this point the need of the hour is for a WG to deliver this in short term, and agree on an evolution strategy long term. Vis a vis the manifests itself, yes this WG should remove it from the scope, as mostly they will be developed in vendors own Github organization e.g. AWS/manifests, IBM/manifests. 

Generic manifests/catalog is being discussed in another thread, and the crux there is should that also be done out of Kubeflow, or is there value in developing that inside Kubeflow - but it's a separate topic.

Jeffwan commented:
@animeshsingh Sounds good. 
I've updated scope and leads in the WG. Please have another check. 

@animeshsingh @cvenets @mameshini @vpavlin @jlewi @Bobgy @yanniszark @paveldournov @theadacty

animeshsingh commented:
/lgtm

Bobgy commented:
/lgtm

RFMVasconcelos commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,34 @@
+
+## Scope
+
+WG Deployment is responsible for tools to deploy apps from the catalog (e.g kfctl).
```
Will this span to other tooling besides kfctl?

RFMVasconcelos commented on wg-deployment/charter.md:
```diff
@@ -0,0 +1,34 @@
+
+#### Code, Binaries and Services
+
+- kfctl (including operator)
```
Is the operator part of `kfctl`?

RFMVasconcelos replied:
@jlewi, given that this is IBM/OpenShift specific, should it be part of upstream Kubeflow?

jlewi commented:
+1 to everything @animeshsingh 

LGTM

@paveldournov and @theadactyl PTAL

jlewi commented:
Thanks @Jeffwan for pushing this over the hump.

RFMVasconcelos commented:
Given the scope:
> - Provide tooling to deploy Kubeflow applications from the catalog.

I believe @knkski would be a great addition as WG lead as he has been working on tooling to deploy Kubeflow applications from the catalog [since Dec 2018](https://github.com/juju-solutions/bundle-kubeflow/graphs/contributors).

knkski commented:
I would be interested in being a part of this WG, provided it's scope is wider than just `kfctl`.

@jlewi: When you say `kfctl is a tool that can be used to define/deploy distributions but it is not a distribution in itself.`, that sounds the same as Charmed Kubeflow, which is tooling around deploying Kubeflow, and does not itself form an opinionated distribution. Given that, it seems like it makes sense to either include tooling like Charmed Kubeflow in the deployment WG, or have this be a `kfctl`-specific WG.

RFMVasconcelos commented:
Hi @jlewi, thank you for the time to elaborate on these thoughts! 

I agree with @animeshsingh's point of starting small with what already exists (`kfctl`) and then allow the WG to define a roadmap. In that sense, I don't think the community can evaluate any addition before the group is formalized, including charms.

I simply think that given @knkski's [track record](https://github.com/juju-solutions/bundle-kubeflow/graphs/contributors) in developing `tooling to deploy Kubeflow applications`, adding him as `wg-lead` will make the group stronger by providing a different, non-kfctl-centric, perspective while defining the 3 months, 6 months, and 1-year roadmap.

Given @animeshsingh's [point](https://github.com/kubeflow/community/pull/434#issuecomment-718934910) that `This can be one WG where the limits on the number of OWNERS can be non-existent. If your org has a track record of contributing back to Kubeflow, join the WG.`, we are happy to help Kubeflow become easy to deploy in any environment.

jlewi commented on wgs.yaml:
```diff
@@ -96,6 +96,76 @@ workinggroups:
+  - name: kfctl
+    owners:
+    - https://raw.githubusercontent.com/kubeflow/kfctl/master/OWNERS
+  - name: manifests
```
@Jeffwan Please remove manifests since we now have a separate WG for this.

Jeffwan replied:
@jlewi forgot removing this part in last change. I just remove it. manifest is totally removed from this PR.

cmhokej replied:
Ok

jlewi commented:
> will make the group stronger by providing a different, non-kfctl-centric, perspective
WGs are open; anyone is welcome to participate and provide perspective.

The rubric and bar for being a lead is much higher. My expectation is that the WG leads are taking responsibility for the success of the projects within the WG. Is @knkski willing to take responsibility for the success of kctl?

PatrickXYS commented:
/lgtm

RFMVasconcelos commented:
@jlewi the current scope is: 
> Provide tooling to deploy Kubeflow applications from the catalog.

@knkski is willing and able to take responsibility for this.

In the context that `kfctl` is a good solution, that took a big amount of effort to produce by the community, I understand your reserves. I also understand the short-term need for `kfctl` to have a backing WG. However, it has been noted that `kfctl` does not solve the general deployment problem and that this group should.

I do not see any good reasons for an open-source community to close doors to contributors with a proven track record on the scope.

PatrickXYS commented:
> I do not see any good reasons for an open-source community to close doors to contributors with a proven track record on the scope.

I think we're saying a not-YES, we're more than happy to see experienced folks to join the community and start contributing. But before he become a Tech Lead in the WG, I assume he should contribute to the Kubeflow Org and/or kfctl repo.

Like what @jlewi or the rest of other WG proposed, the WG Tech Lead list can be extended once we have desired candidate which proves his contribution within last 6 months / 1 year.

Besides that, even if folks are not Lead of the WG-deployment, they can still join our meeting and propose their idea for development.

jlewi commented:
Thanks @Jeffwan

Jeffwan commented:
> In the context that kfctl is a good solution, that took a big amount of effort to produce by the community, I understand your reserves. I also understand the short-term need for kfctl to have a backing WG. However, it has been noted that kfctl does not solve the general deployment problem and that this group should.

> I do not see any good reasons for an open-source community to close doors to contributors with a proven track record on the scope.

@RFMVasconcelos 
Door is alway open for sure. We definitely want to have more contributors. This WG is not limited to `kfctl` project itself but `tooling to deploy Kubeflow applications from the catalog.` as Jeremy said. Currently, all the folks listed here are kfctl contributors. We have not extended to other different tools. Please forgive that we spend too much time on the scope discussion and I really like to close this and have more discussion inside WG later. No worries on the proposed wg leads, this will be changed on a regular base as @PatrickXYS said, we really want to invite more folks to lead the direction and come up new ideas to help this area grow.  I do see juju demo last time in the community meeting, this is excellent and we also like to know how it can improve the experience of existing Kubeflow users. Hope you guys can also check existing kfctl solution since this is mostly used tool in the community.

RFMVasconcelos commented:
Hi @PatrickXYS, @Jeffwan, @jlewi, 

Thank you for the clarification, for being pacient, and for being open to future collaboration! 

Do let us know how we can contribute in the future. 

@Jeffwan, we will explore `kfctl` and are happy to demo `juju deploy kubeflow` anytime :)

animeshsingh commented:
/lgtm

Let's close this folks, open now for more than two months.  @paveldournov @theadactyl @jlewi

PatrickXYS commented:
Yeah let's get this merged 

/lgtm

theadactyl commented:
Thanks everyone for the discussion and working to get this charter together. 

/lgtm
/approve

PatrickXYS commented:
/cc @Jeffwan 
Can you rebase to pass the conflict issue?

Jeffwan commented:
/lgtm
